File: android/app/src/main/java/dr/achim/sleep_timer/AlarmService.java
package dr.achim.sleep_timer;

import android.app.AlarmManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.os.Build;
import android.os.IBinder;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import dr.achim.sleep_timer.Messages.*;

import java.util.Calendar;
import java.util.HashMap;

import static dr.achim.sleep_timer.NotificationReceiver.*;

public class AlarmService extends Service {
    private static final String TAG = AlarmService.class.toString();
    private static boolean isRunning;

    public static final String ACTION_START = "ACTION_START";
    public static final String ACTION_STOP = "ACTION_STOP";
    private static final int REQUEST_CODE_ALARM = 700;
    private PendingIntent pendingAlarmIntent;

    private SensorManager sensorManager;
    private Sensor accelerometer;
    private ShakeDetector shakeDetector;

    @Override
    public void onCreate() {
        super.onCreate();

        initNotificationChannel();
        initShakeDetector();

        final Notification notification = new NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID).build();
        startForeground(NOTIFICATION_ID, notification);
        Log.wtf(TAG, "AlarmService onCreate");
    }

    private void initNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            final int importance = NotificationManager.IMPORTANCE_DEFAULT;
            final NotificationChannel notificationChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, "Active timer", importance);
            notificationChannel.setDescription("Notify about running or pausing timers");
            NotificationManagerCompat.from(this).createNotificationChannel(notificationChannel);
        }
    }

    private void initShakeDetector() {
        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        accelerometer = sensorManager
                .getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        shakeDetector = new ShakeDetector();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        isRunning = true;

        Log.wtf(TAG, intent.getAction());
        switch (intent.getAction()) {
            case ACTION_START:
                final HashMap map = (HashMap) intent.getSerializableExtra(KEY_SHOW_NOTIFICATION);
                final TimeNotificationRequest request = TimeNotificationRequest.fromMap(map);
                _startAlarm(request);

                final Intent showRunningIntent = new Intent(this, NotificationReceiver.class);
                showRunningIntent.setAction(ACTION_SHOW_RUNNING);
                showRunningIntent.putExtra(KEY_SHOW_NOTIFICATION, map);
                sendBroadcast(showRunningIntent);
                break;
            case ACTION_STOP:
                stopForeground(true);
                stopSelf();
            default:
                if(isRunning()) {
                    isRunning = false;
                    stopForeground(true);
                    stopSelf();
                }
                break;
        }

        return START_STICKY;
    }

    private void _startAlarm(TimeNotificationRequest request) {
        final Intent alarmIntent = new Intent(this, AlarmReceiver.class);
        alarmIntent.putExtra(NotificationReceiver.KEY_TIMER_ID, request.getTimerId());
        pendingAlarmIntent = PendingIntent.getBroadcast(this, REQUEST_CODE_ALARM, alarmIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        final AlarmManager manager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        manager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, Calendar.getInstance().getTimeInMillis() + request.getRemainingTime() * 1000, pendingAlarmIntent);

        shakeDetector.setOnShakeListener(count -> {
            final Intent intent = new Intent(this, NotificationActionReceiver.class);
            intent.setAction(ACTION_EXTEND);

            final ExtendTimeResponse response = new ExtendTimeResponse();
            response.setTimerId(request.getTimerId());
            intent.putExtra(KEY_EXTEND_RESPONSE, response.toMap());

            sendBroadcast(intent);

            Vibrator vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibrator.vibrate(VibrationEffect.createOneShot(500, VibrationEffect.DEFAULT_AMPLITUDE));
            } else {
                vibrator.vibrate(500);
            }
        });
        sensorManager.registerListener(shakeDetector, accelerometer, SensorManager.SENSOR_DELAY_UI);
    }

    private void _stopAlarm() {
        if(pendingAlarmIntent != null) {
            final AlarmManager manager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
            manager.cancel(pendingAlarmIntent);
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    public void onDestroy() {
        Log.wtf(TAG, "onDestroy Service");
        _stopAlarm();
        sensorManager.unregisterListener(shakeDetector);
        isRunning = false;
    }
    public static boolean isRunning() {
        return isRunning;
    }
}

File: android/app/src/main/kotlin/dr/achim/sleep_timer/AlarmService.kt
package dr.achim.sleep_timer

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ServiceInfo
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.*
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import dr.achim.sleep_timer.Messages.ExtendTimeRequest
import dr.achim.sleep_timer.Messages.RunningNotificationRequest
import dr.achim.sleep_timer.ShakeDetector.OnShakeListener
import java.util.*
import kotlin.math.max

class AlarmService : Service() {

    companion object {
        private val TAG = AlarmService::class.java.toString()
        var isRunning = false
        const val ACTION_START = "ACTION_START"
        const val ACTION_TOGGLE_EXTEND_BY_SHAKE = "ACTION_TOGGLE_EXTEND_BY_SHAKE"
        const val KEY_ENABLE_EXTEND_BY_SHAKE = "KEY_ENABLE_EXTEND_BY_SHAKE"
        private const val REQUEST_CODE_ALARM = 700
    }

    private var timerId: String? = null
    private var timer: Timer = Timer()
    private var notification: Notification? = null

    private var sensorManager: SensorManager? = null
    private var shakeDetector: ShakeDetector? = null
    private var accelerometer: Sensor? = null
    private var pendingAlarmIntent: PendingIntent? = null

    private fun startForeground() {
        try {
            if (notification == null) {
                initNotificationChannel()
                notification =
                    NotificationCompat.Builder(this, NotificationReceiver.NOTIFICATION_CHANNEL_ID)
                        .build()
            }

            notification?.let {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    startForeground(
                        NotificationReceiver.NOTIFICATION_ID,
                        it,
                        ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
                    )
                } else {
                    startForeground(NotificationReceiver.NOTIFICATION_ID, it)
                }
            }

            if (packageManager.hasSystemFeature(PackageManager.FEATURE_SENSOR_ACCELEROMETER)) {
                initShakeDetector()
            }

        } catch (e: Exception) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
                && e is ForegroundServiceStartNotAllowedException
            ) {
                Log.d(TAG, "App not in a valid state to start foreground service")
            }
            Log.d(TAG, e.localizedMessage ?: e.message ?: e.toString())
        }
    }

    private fun initNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_LOW
            val notificationChannel = NotificationChannel(
                NotificationReceiver.NOTIFICATION_CHANNEL_ID,
                "Active timer",
                importance
            )
            notificationChannel.description = "Notify about running or pausing timers"
            NotificationManagerCompat.from(this).createNotificationChannel(notificationChannel)
        }
    }

    private fun initShakeDetector() {
        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
        accelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        if (accelerometer != null) {
            shakeDetector = ShakeDetector()
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        startForeground()
        isRunning = true
        Log.d(TAG, "intent action: ${intent?.action}")

        when (intent?.action) {
            ACTION_START -> {
                val map =
                    intent.getSerializableExtra(NotificationReceiver.KEY_SHOW_NOTIFICATION) as ArrayList<Any>?

                if (map != null) {
                    val request: RunningNotificationRequest =
                        RunningNotificationRequest.fromList(map)
                    startAlarm(request)

                    timer.schedule(object : TimerTask() {
                        override fun run() {
                            val response = Messages.CountDownRequest().apply {
                                timerId = request.timerId

                                request.remainingTime = (request.remainingTime ?: 0L) - 1L
                                newTime = max(request.remainingTime!!, 0L)
                            }

                            val countDownIntent = Intent(
                                applicationContext,
                                NotificationActionReceiver::class.java
                            ).apply {
                                action = NotificationReceiver.ACTION_COUNTDOWN
                                putExtra(
                                    NotificationReceiver.KEY_COUNTDOWN_REQUEST,
                                    response.toList()
                                )
                            }
                            sendBroadcast(countDownIntent)

                            val showRunningIntent =
                                Intent(applicationContext, NotificationReceiver::class.java).apply {
                                    action = NotificationReceiver.ACTION_SHOW_RUNNING
                                    putExtra(
                                        NotificationReceiver.KEY_SHOW_NOTIFICATION,
                                        request.toList()
                                    )
                                }
                            sendBroadcast(showRunningIntent)
                        }
                    }, 0, 1000)
                }
            }

            ACTION_TOGGLE_EXTEND_BY_SHAKE -> {
                val enable = intent.getBooleanExtra(KEY_ENABLE_EXTEND_BY_SHAKE, false)

                if (isRunning) {
                    if (enable) {
                        startShakeListener(timerId)
                    } else {
                        sensorManager?.unregisterListener(shakeDetector)
                    }
                }
            }

            else -> {
                Log.d(TAG, "Stopping service due to ${intent?.action}")
                isRunning = false
                stopForeground(STOP_FOREGROUND_REMOVE)
                stopSelf()
            }
        }
        return START_STICKY
    }

    private fun startAlarm(request: RunningNotificationRequest) {
        Log.d(TAG, "startAlarm")
        val alarmIntent = Intent(this, AlarmReceiver::class.java)
        alarmIntent.putExtra(NotificationReceiver.KEY_TIMER_ID, request.timerId)

        val manager = getSystemService(Context.ALARM_SERVICE) as AlarmManager

        pendingAlarmIntent = PendingIntent.getBroadcast(
            this,
            REQUEST_CODE_ALARM,
            alarmIntent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
        pendingAlarmIntent?.let { intent ->
            val triggerInMillis =
                Calendar.getInstance().timeInMillis + request.remainingTime!! * 1000
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S || manager.canScheduleExactAlarms()) {
                manager.setExactAndAllowWhileIdle(
                    AlarmManager.RTC_WAKEUP,
                    triggerInMillis,
                    intent,
                )
            } else {
                manager.setAndAllowWhileIdle(
                    AlarmManager.RTC_WAKEUP,
                    triggerInMillis,
                    intent,
                )
            }
        }

        timerId = request.timerId
        if (request.shakeToExtend == true) {
            startShakeListener(timerId)
        }
    }

    private fun startShakeListener(timerId: String?) {
        shakeDetector?.setOnShakeListener(object : OnShakeListener {
            override fun onShake(count: Int) {
                val response = ExtendTimeRequest()
                response.timerId = timerId ?: this@AlarmService.timerId

                val intent =
                    Intent(applicationContext, NotificationActionReceiver::class.java).apply {
                        action = NotificationReceiver.ACTION_EXTEND
                        putExtra(NotificationReceiver.KEY_EXTEND_RESPONSE, response.toList())
                    }
                sendBroadcast(intent)

                val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    val vibratorManager =
                        getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                    vibratorManager.defaultVibrator
                } else {
                    @Suppress("DEPRECATION")
                    getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
                }

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibrator.vibrate(
                        VibrationEffect.createOneShot(
                            500,
                            VibrationEffect.DEFAULT_AMPLITUDE
                        )
                    )
                } else {
                    @Suppress("DEPRECATION")
                    vibrator.vibrate(500)
                }
            }
        })
        sensorManager?.registerListener(shakeDetector, accelerometer, SensorManager.SENSOR_DELAY_UI)
    }

    private fun stopAlarm() {
        Log.d(TAG, "stopAlarm")
        pendingAlarmIntent?.let {
            val manager = getSystemService(ALARM_SERVICE) as AlarmManager
            manager.cancel(it)
        }
        timer.cancel()
        timerId = null
        sensorManager?.unregisterListener(shakeDetector)
        isRunning = false
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy")
        stopAlarm()
        super.onDestroy()
    }

}